pkg/api/message/v3/service.go

```diff
@@ -45,11 +48,50 @@ func (s *Service) Close() {
 }
 
 func (s *Service) RegisterInstallation(ctx context.Context, req *proto.RegisterInstallationRequest) (*proto.RegisterInstallationResponse, error) {
-	return nil, status.Errorf(codes.Unimplemented, "unimplemented")
+	results, err := s.validationService.ValidateKeyPackages(ctx, [][]byte{req.LastResortKeyPackage.KeyPackageTlsSerialized})
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "invalid identity: %s", err)
+	}
+	if len(results) != 1 {
+		return nil, status.Errorf(codes.Internal, "unexpected number of results: %d", len(results))
+	}
+
+	installationId := results[0].InstallationId
+	walletAddress := results[0].WalletAddress
+
+	err = s.mlsStore.CreateInstallation(ctx, installationId, walletAddress, req.LastResortKeyPackage.KeyPackageTlsSerialized)
+	if err != nil {
+		return nil, err
+	}
+
+	return &proto.RegisterInstallationResponse{
+		InstallationId: installationId,
+	}, nil
 }
 
 func (s *Service) ConsumeKeyPackages(ctx context.Context, req *proto.ConsumeKeyPackagesRequest) (*proto.ConsumeKeyPackagesResponse, error) {
-	return nil, status.Errorf(codes.Unimplemented, "unimplemented")
+	ids := req.InstallationIds
+	keyPackages, err := s.mlsStore.ConsumeKeyPackages(ctx, ids)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to consume key packages: %s", err)
+	}
+
+	resPackages := make([]*proto.ConsumeKeyPackagesResponse_KeyPackage, len(keyPackages))
+	for _, keyPackage := range keyPackages {
+		// Return the key packages in the original order
+		targetIndex := indexOf(keyPackage.InstallationId, ids)
```

Comment:
Want to build a map of id to index above this loop and do the O(1) lookup here instead of this O(n) indexOf?

Reply:
👍🏼